// @yidun/snaker/dist/lib/userStore.js
"use strict";
exports.__esModule = true;
var userStore = {
    userData: null,
    name: location.hostname + "_snaker",
    init: function () {
        if (!userStore.userData) {
            try {
                userStore.userData = document.createElement('INPUT');
                userStore.userData.type = 'hidden';
                userStore.userData.style.display = 'none';
                userStore.userData.addBehavior('#default#userData');
                userStore.userData && document.body.appendChild(userStore.userData);
                var expires = new Date();
                expires.setDate(expires.getDate() + 365);
                userStore.userData.expires = expires.toUTCString();
            }
            catch (error) {
                console.log('userData is disabled!');
                return false;
            }
        }
        return true;
    },
    setItem: function (key, value) {
        if (userStore.init() && userStore.userData) {
            userStore.userData.load(userStore.name);
            userStore.userData.setAttribute(key, value);
            userStore.userData.save(userStore.name);
        }
    },
    getItem: function (key) {
        if (userStore.init() && userStore.userData) {
            userStore.userData.load(userStore.name);
            return userStore.userData.getAttribute(key) || '';
        }
        return '';
    },
    removeItem: function (key) {
        if (userStore.init() && userStore.userData) {
            userStore.userData.load(userStore.name);
            userStore.userData.removeAttribute(key);
            userStore.userData.save(userStore.name);
        }
    }
};
exports["default"] = userStore;
