// @yidun/snaker/dist/lib/plugins/vue.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
exports.__esModule = true;
var index_1 = require("../index");
var webClick_1 = require("./webClick");
var utils_1 = require("./utils");
exports["default"] = {
    install: function (Vue, pluginOptions) {
        var router = pluginOptions.router, _a = pluginOptions.webClick, webClick = _a === void 0 ? false : _a, options = __rest(pluginOptions, ["router", "webClick"]);
        var snaker = new index_1["default"](options);
        var report = {
            binding: function (event) { },
            add: function (el, value) {
                this.binding = function () {
                    snaker.trackAsync('event', 'click', value);
                };
                el.addEventListener('click', this.binding);
            },
            remove: function (el) {
                el.removeEventListener('click', this.binding);
            }
        };
        Vue.directive('sk', {
            bind: function (el, binding) {
                report.add(el, binding.value);
            },
            unbind: function (el) {
                report.remove(el);
            }
        });
        Vue.prototype.$snaker = snaker;
        if (router) {
            router.afterEach(function (to, from) {
                snaker.trackAsync('pageView', to.path, from.path);
            });
        }
        if (webClick) {
            var instance = webClick_1["default"](snaker, function (ev) {
                var el = ev.target;
                var vpath = utils_1.getElementVPath(el);
                return {
                    vpath: vpath
                };
            });
            instance.start();
        }
    }
};
